darts/timeseries.py

```diff
@@ -496,6 +497,119 @@ def from_dataframe(
+            :meth:`_fill_missing_dates() <TimeSeries._fill_missing_dates>` for more info.
+        freq
+            Optionally, a string representing the frequency of the Pandas DateTimeIndex. This is useful in order to
+            fill in missing values if some dates are missing and `fill_missing_dates` is set to `True`.
+        fillna_value
+            Optionally, a numeric value to fill missing values (NaNs) with.
+
+        Returns
+        -------
+        TimeSeries
+            A univariate or multivariate deterministic TimeSeries constructed from the inputs.
+        """
+        group_cols = [group_cols] if not isinstance(group_cols, list) else group_cols
+        if static_cols is not None:
+            static_cols = (
+                [static_cols] if not isinstance(static_cols, list) else static_cols
+            )
+        else:
+            static_cols = []
+        static_cov_cols = group_cols + static_cols
+
+        # split df by groups, and store group values and static values (static covariates)
+        splits = []
+        for static_cov_vals, group in df.groupby(group_cols):
+            static_cov_vals = (
+                (static_cov_vals,)
+                if not isinstance(static_cov_vals, tuple)
+                else static_cov_vals
+            )
+            # check that for each group there is only one unique value per column in `static_cols`
+            if static_cols:
+                static_cols_valid = [
+                    len(group[col].unique()) == 1 for col in static_cols
+                ]
+                if not all(static_cols_valid):
+                    # encountered performance issues when evaluating the error message from below in every
+                    # iteration with `raise_if_not(all(static_cols_valid), message, logger)`
+                    invalid_cols = [
+                        static_col
+                        for static_col, is_valid in zip(static_cols, static_cols_valid)
+                        if not is_valid
+                    ]
+                    raise_if(
+                        True,
+                        f"Encountered more than one unique value in group {group} for given static columns: "
+                        f"{invalid_cols}.",
+                        logger,
+                    )
+                # add the static covariates to the group values
+                static_cov_vals += tuple(group[static_cols].values[0])
+            # store static covariate Series and group DataFrame (without static cov columns)
+            splits.append(
+                (
+                    pd.DataFrame(static_cov_vals, index=static_cov_cols),
+                    group.drop(columns=static_cov_cols),
+                )
+            )
+
+        # create a list with multiple TimeSeries and add static covariates
+        return [
+            TimeSeries.from_dataframe(
```

Comment:
A possibility could be to directly accept `static_covariates` in all (or most of) our `from_XXX` methods. This way we would get a chance to build the DataArray directly right, containing the covariates, and it'd avoid copying it later on in in the method to set the static covariates (see my other comments there to keep TimeSeries immutable).

But that's just a thought - it might be overkill and avoiding this copy not really needed for now.